src/dspbase.jl

```diff
@@ -118,75 +118,319 @@ function deconv(b::StridedVector{T}, a::StridedVector{T}) where T
+        block_out_region = CartesianIndices(
+            UnitRange.(data_offset .+ out_start, data_stop .+ out_start .- 1)
+        )
+        copyto!(out, block_out_region, tdbuff, valid_buff_region)
+    end
+
+    out
 end
 
-function _conv(
-    u::AbstractArray{T, N}, v::AbstractArray{T, N}, paddims
+function _conv_kern_fft!(
+    out, u::AbstractArray{T, N}, v::AbstractArray{T, N}, su, sv, outsize, nffts
 ) where {T<:Real, N}
-    padded = _zeropad(u, paddims)
+    padded = _zeropad(u, nffts)
     p = plan_rfft(padded)
     uf = p * padded
     _zeropad!(padded, v)
     vf = p * padded
     uf .*= vf
-    irfft(uf, paddims[1])
+    raw_out = irfft(uf, nffts[1])
+    copyto!(out, CartesianIndices(out), raw_out,
+            CartesianIndices(UnitRange.(1, outsize)))
 end
-function _conv(u, v, paddims)
-    upad = _zeropad(u, paddims)
-    vpad = _zeropad(v, paddims)
+function _conv_kern_fft!(out, u, v, su, sv, outsize, nffts)
+    upad = _zeropad(u, nffts)
+    vpad = _zeropad(v, nffts)
     p! = plan_fft!(upad)
     p! * upad # Operates in place on upad
     p! * vpad
     upad .*= vpad
     ifft!(upad)
+    copyto!(out, CartesianIndices(out), upad,
+            CartesianIndices(UnitRange.(1, outsize)))
 end
 
-function _conv_clip!(
-    y::AbstractVector,
-    minpad,
-    ::NTuple{<:Any, Base.OneTo{Int}},
-    ::NTuple{<:Any, Base.OneTo{Int}}
-)
-    sizehint!(resize!(y, minpad[1]), minpad[1])
-end
-function _conv_clip!(
-    y::AbstractArray,
-    minpad,
-    ::NTuple{<:Any, Base.OneTo{Int}},
-    ::NTuple{<:Any, Base.OneTo{Int}}
-)
-    y[CartesianIndices(minpad)]
+function _conv_fft!(out, u, v, su, sv, outsize)
+    os_nffts = map((nu, nv)-> optimalfftfiltlength(nu, nv), su, sv)
+    if any(os_nffts .< outsize)
+        _conv_kern_os!(out, u, v, su, sv, outsize, os_nffts)
+    else
+        nffts = nextfastfft(outsize)
+        _conv_kern_fft!(out, u, v, su, sv, outsize, nffts)
+    end
 end
+
+
 # For arrays with weird offsets
-function _conv_clip!(y::AbstractArray, minpad, axesu, axesv)
+function _conv_similar(u, outsize, axesu, axesv)
     out_offsets = first.(axesu) .+ first.(axesv)
-    out_axes = UnitRange.(out_offsets, out_offsets .+ minpad .- 1)
-    out = similar(y, out_axes)
-    copyto!(out, CartesianIndices(out), y, CartesianIndices(UnitRange.(1, minpad)))
+    out_axes = UnitRange.(out_offsets, out_offsets .+ outsize .- 1)
+    similar(u, out_axes)
+end
+function _conv_similar(
+    u, outsize, ::NTuple{<:Any, Base.OneTo{Int}}, ::NTuple{<:Any, Base.OneTo{Int}}
+)
+    similar(u, outsize)
 end
+_conv_similar(u, v, outsize) = _conv_similar(u, outsize, axes(u), axes(v))
 
+# Does convolution, will not switch argument order
+function _conv!(out, u, v, su, sv, outsize)
+    # TODO: Add spatial / time domain algorithm
+    _conv_fft!(out, u, v, su, sv, outsize)
+end
+
+# Does convolution, will not switch argument order
+function _conv(u, v, su, sv)
+    outsize = su .+ sv .- 1
+    out = _conv_similar(u, v, outsize)
```

Comment:
Maybe now that it's only used once, but it feels like an implementation detail that the caller shouldn't need to worry about.

Reply:
Given that the caller is already an internal function my preference is for less abstraction so it's easier to follow the code, but I'm happy to agree to disagree on this one.

Reply:
Maybe it's just a poorly named function. I think it feels more natural to have an interface like `_conv_similar(u, v, outsize)` because it seems similar to `similar` and makes a new array with appropriate axes to store the convolution result between `u` and `v`, but a interface like `_conv_similar(u, outsize, axes_u, axes_v)` doesn't make too much sense to me, even if that's what the next step in the process is.